### DISCUSSION TITLE
[enhancement] allow SRAM only programming

### DISCUSSION THREAD

**Discussion by chriscamacho** [FPGAwars/icestudio#854] [Ideas]
allow uploading only to fpga sram and not to flash, I regularly use iceprog -S as it is *much* faster especially while testing, obviously once tested you'd want to revert to programming flash...

**Comment by Jesus89** [comment]
Icestudio uses [apio](https://github.com/FPGAwars/apio) as backend. Any enhancement in the execution of the toolchain must be implemented first in apio.

Then it can be implemented in Icestudio. Since Icestudio supports multiple boards with different programmers, do you have any preference in how to implement the UX/UI of this feature?

**Comment by chriscamacho** [comment]
in my makefile for icestorm I have two targets prog and prog-ram

in a similar manner in the menu you could have "upload" and "upload to sram" it should be noted on my lattice breakout board, you have to remove one jumper and rotate a block of two other jumpers before you can upload to sram and this then prevents upload to flash...

for usability reasons a better way might be to have a project setting say a check box "upload to sram instead of flash" (for the boards that can do it) and then only have an upload menu item. By default upload should go to flash 

The circuit diagrams for the lattice ice40hx8k breakout board are readily available and it might be worth looking at how these jumpers work as some other boards might be able to support this functionality with a minor "tweak" - for all I know some others might be able to do it already

**Comment by Jesus89** [comment]
Great. I like the "Upload to SRAM" checkbox. We want to see if it is a project setting, a user setting and where to put this setting: Tools, Edit, ...

Thanks for the suggestion. It will be implemented in future releases.

**Comment by chriscamacho** [comment]
I'd have thought a project setting, but also consider a pop up when selected to warn the user to change the jumpers, I know after leaving my board for a while I couldn't program flash forgetting about the jumper settings but don't know if there is anyway to infer how the physical jumpers are set ?

**Comment by nmz787** [comment]
I just started trying icestudio tonight (linux nightly).... after cutting the traces on the icebreaker and successfully getting SRAM upload working on the command-line. Is this feature really still unsupported, or has this issue simply not been updated?

**Comment by nmz787** [comment]
this `apio` PDF says their upload command supports `-s` https://readthedocs.org/projects/apiodoc/downloads/pdf/latest/

```
-s, --sram
Perform SRAM programming. Only available for iceprog compatible boards.
```

how can I quickly hack this to work? Clicking upload has now frozen Icestudio and there's no **cancel** button.

edit: I forcefully restarted icestudio and found that even though I'd clicked verify, build, and then upload, my work hadn't saved and I lost some progress.

**Comment by nmz787** [comment]
ok, after a little digging, I found the `/tmp/icestudio` prefix is where active projects are built. Since I only had one project open, after `Build`ing I was able to just `cd` into the latest directory... i.e. 
```
cd `ls -tr /tmp | grep icestudio | tail -1`
```

then `iceprog -S hardware.bin` to shove it into the icebreaker's SRAM

**Comment by Codedial06** [comment]
Hello, i tried to program into SRAM but I don't know how to change the apio settings, can somebody help me please I'm using Win10

**Comment by cavearr** [comment]
Hello! I am reorganizing the issues, trying to give more visibility and be able to better organize and prioritize the work and my roadmap (I will publish it very soon), I am converting some issues that are **new feature requests** as **ideas** on Github **discussions**.

Thanks for understanding.
